### ISSUE TITLE
False positive with ? in `clippy::if_then_some_else_none`

### ISSUE THREAD

**Issue by Awpteamoose** [rust-lang/rust-clippy#11394]
### Summary

Probably related to #7870, if I remove `let _ = ` then the warning disappears

### Lint Name

clippy::if_then_some_else_none

### Reproducer

I tried this code:

```rust
#![warn(clippy::if_then_some_else_none)]

fn f(b: bool, v: Result<(), ()>) -> Result<(), ()> {
    let x = if b {
        let _ = v?;
        Some(())
    } else {
        None
    };

    Ok(())
}
```

I saw this happen:

```
warning: this could be simplified with `bool::then`
```

I expected to see this happen: No warning

### Version

```text
rustc 1.73.0-nightly (1b198b3a1 2023-08-13)
binary: rustc
commit-hash: 1b198b3a196442e14fb06978166ab46a4618d131
commit-date: 2023-08-13
host: x86_64-pc-windows-msvc
release: 1.73.0-nightly
LLVM version: 17.0.0
```
